[1단계 방탈출 결제 / 배포] 아마(유석현) 미션 제출합니다.

Astro-Yu opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://github.com/woowacourse/spring-roomescape-payment/pull/197/files/9ee9f1f4da158067ac6e7f550951e135601f55c6..42104908114df3b261f3fa6f93db37b837ced54d)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 프롤로그의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/4423

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

## 안녕하세요 이스트! 다시 만나서 반가워요
이번 미션의 변경 커밋: [링크](https://github.com/woowacourse/spring-roomescape-payment/pull/197/files/9ee9f1f4da158067ac6e7f550951e135601f55c6..5023ee573d88e415e28f7d4bea2fed65d32a35e8) 6월 4일 업데이트
이번 미션에서 외부 API를 사용해보는 경험을 하면서 무엇에 집중해야 하나 고민했습니다.
코드를 작성하면서 생각한 부분은 다음과 같아요.
### PaymentClientXXX 의 적절한 패키지 위치는 어디인가.
처음엔 결제 서비스를 제공한다고 생각해 당연히 service 패키지에 위치시켰습니다만, 몇가지 이유로 presentation 패키지로 이동시켰습니다.
1. service에는 방탈출과 관련된 도메인 로직이 들어있는데, 결제가 방탈출 로직과 관련 없다고 생각했음.
2. 결제 기능은 단순히 요청을 받아서 응답을 반환한다고 생각해 컨트롤러가 있는 presentation에 위치하는게 적절하다고 생각했음.

### 외부 API에서 오는 예외 처리는 어떻게 해야할까.
토스 페이먼츠에서 오는 대부분의 예외를 그대로 내보내고 있지만, 몇몇 예외의 경우 사용자에게 내보내는게 적절하지 않다고 판단되어 필터링 했습니다. 이 중 400번대 에러를 우리 방탈출 도메인에선 서버 에러라고 생각해 500번 에러로 변경하였습니다.
해당 코드 위치는 코멘트로 남기겠습니다. 적절한 예외 처리 방식인지 궁금하고, 제가 생각하지 못한 부분이 있다면 말씀해주세요!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

Astro-Yu committed:
feat: first commit

kwonkeonhyeong committed:
docs: 기능 구현 사항 작성

Astro-Yu committed:
feat: 예약화면 결제ui 추가

Astro-Yu committed:
feat: 결제 config 등록

kwonkeonhyeong committed:
feat: 결제 승인 요청 구현

kwonkeonhyeong committed:
feat: 결제 승인 예외 처리 기능 구현

Astro-Yu committed:
feat: connection, request timeout 설정

Astro-Yu committed:
refactor: PaymentClientService 패키지 이동

kwonkeonhyeong committed:
test: Reservation Test 수정

Astro-Yu committed:
test: PaymentClientController 테스트 구현

kwonkeonhyeong committed:
feat: 필터링 된 예외 추가

Astro-Yu committed:
test: 필터링된 예외 테스트 추가

kwonkeonhyeong committed:
chore: 필요 없는 테스트 제거

kwonkeonhyeong committed:
fix: PaymentClientController.java @Component 애노테이션 적용

EastHShin commented on README.md:
```diff
@@ -0,0 +1,17 @@
+    - [ ] tosspay 사용
+
+- [ ] 결제 승인 API 호출에 실패 한 경우, 안전하게 에러를 핸들링 하세요.
+    - [ ] 사용자는 예약 실패 시, 결제 실패 사유를 알 수 있어야 합니다.
```
구현된 기능은 체크박스를 채워주시면 좋겠습니다 :)

Astro-Yu replied:
헉 죄송합니다 ㅠㅠ

EastHShin commented on src/main/java/roomescape/config/ClientConfig.java:
```diff
@@ -0,0 +1,25 @@
+    public RestClient paymentClient() {
+        HttpComponentsClientHttpRequestFactory factory = new HttpComponentsClientHttpRequestFactory();
+        factory.setConnectionRequestTimeout(5_000);
+        factory.setReadTimeout(5_000);
```
connectTimeout, readTimeout 을 설정해주셨군요!
이 둘의 차이는 무엇이고, 어떤 기준으로 5초를 설정해주셨나요?

Astro-Yu replied:
### Connection Timeout
클라이언트가 서버에 소켓 연결을 시도할 때의 타임아웃. 서버 응답 불가 상황에서 무한 연결시도를 방지합니다.

### ConnectionRequestTimeout
커넥션 풀에서 사용 가능한 커넥션을 얻기 까지의 대기시간. 모든 커넥션이 사용중일때, 대기 시간을 설정합니다.

###  ReadTimeout
클라이언트가 서버의 응답을 기다리는 시간. 네트워크가 불안정하거나, 서버에서 처리 시간이 오래 걸리는 경우를 방지합니다.

5초로 설정한 것은 해당 개념들을 처음 접해봐서, [예제](https://www.baeldung.com/rest-template#configure-timeout) 에 있는 시간을 그대로 사용했습니다.
토스 페이먼츠 문서에서는 ReadTimeout의 경우 30~60초를 설정해주라고 하니 그대로 수정하겠습니다.

EastHShin replied:
음, confirm 과 같은 API 에서 30\~60 초는 너무 긴 것 같아요. 
토스 페이먼츠 문서에서 30\~60 초를 권장한 것은, 신용카드 결제와 같이 처리시간이 오래 걸리는 API에 대해 그렇게 안내한 것으로 보이네요.
서버에서 30\~60초와 같이 길게 timeout을 잡았을 경우에 요청이 몰리게 된다면 어떤 일이 발생할 수 있을까요?

Astro-Yu replied:
토스 문서를 더 살펴보니 자동 결제를 하는 상황에서, 자동 결제 승인의 최대 시간이 30초라 30 ~ 60초를 추천한다고 합니다.
현재 제가 사용하는 결제 방식은 일반 방식이니, 시간을 줄이겠습니다.
### 긴 타임아웃을 가졌을 때, 요청이 몰리게 되면 발생하는 일
1. 모든 요청이 최악의 경우(외부 요청마다 30초가 소요됨)일 때, 하나의 스레드가 30초를 점유함.
2. 최대 스레드를 200개, 초당 요청이 200개라고 했을 때, 1초만에 모든 스레드를 점유하게 됨.
3. 다음 요청부터는 다음 스레드가 반환될 때 까지 30초를 기다려야함.
4. 새 요청 처리 불가.

### 해결방안.
- readTimeout 시간을 낮춘다.
- 요청마다 다른 readTimeout을 적용한다.

EastHShin commented on src/main/java/roomescape/config/ClientConfig.java:
```diff
@@ -0,0 +1,25 @@
+        return RestClient.builder()
+                .baseUrl("https://api.tosspayments.com")
+                .defaultHeader("Authorization", String.format("%s %s", "Basic", Base64.getEncoder()
+                        .encodeToString("test_gsk_docs_OaPz8L5KdmQXkzRz3y47BMw6:".getBytes())))
```
secretKey는 상수로 추출해주시면 좋겠네요~

EastHShin commented on src/main/java/roomescape/exception/PaymentException.java:
```diff
@@ -0,0 +1,10 @@
+package roomescape.exception;
+
+import org.springframework.http.HttpStatus;
```
사용되지 않는군요~

EastHShin commented on src/main/java/roomescape/exception/PaymentException.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpStatusCode;
+
+public class PaymentException extends CustomException {
```
CustomException 을 상속하신 이유가 있나요? 예외 핸들러를 보면 CustomException, PaymentException이 각각 핸들링 되고 있는데, 그렇다면 RuntimeException 을 상속해주어도 될 것 같네요

Astro-Yu replied:
생각해보니 그렇네요. 차라리 예외 핸들러에서 PaymentException을 제거하는 편이 PaymentException에서 상태코드를 계속 가져갈 수 있을 것 같습니다. 해당 방향으로 수정해보겠습니다.

EastHShin commented on src/main/java/roomescape/exception/PaymentErrorResponse.java:
```diff
@@ -0,0 +1,4 @@
+package roomescape.exception;
+
+public record PaymentErrorResponse(String code, String message) {
```
http 요청의 response 에 대한 정의인데, exception 패키지에 있는 것이 어색하진 않으신가요?

Astro-Yu replied:
일단 해당 코드는 현재 아무 곳에서도 사용되고 있지 않아서 삭제하겠습니다.
리뷰를 받고 보니 dto 패키지의 response에 위치하는 편이 적절하다는 생각이 들면서도, 해당 패키지 내에는 "실제로" 반환되는 response 값만 모여있는데, `PaymentErrorResponse`은 실제 반환값으로 사용한다기 보다는 내부적으로 에러 코드를 활용하는 역할로만 사용했을 것 같습니다.(최종적으로 onStatus를 사용해서 실제 `PaymentErrorResponse`가 활용되지 않았습니다.) 그래서 exception 내부에 있는 `ErrorResponse` 처럼 해당 위치에 넣었습니다. `PaymentErrorResponse`라는 네이밍이 적절하지 않았던 것 같기도 하네요. 어떻게 생각하시나요?

EastHShin replied:
PaymentErrorResponse 라고 하면 결제 API에서 응답받은 것을 사용한다고 생각하게 되는 것 같네요. 말씀 주신 것 처럼 네이밍이 적절하지 않았던 것 같아요~
exception 패키지에는 정의한 예외 클래스를 넣어주는 것이 적절하다고 생각합니다.

EastHShin commented on src/main/java/roomescape/presentation/PaymentClientController.java:
```diff
@@ -0,0 +1,47 @@
+                        throw new FilteredPaymentException();
+                    }
+                    throw new PaymentException(message, response.getStatusCode());
+                })
```
responseErrorHandler 를 상속한 클래스를 정의해 핸들링 로직을 분리해보는 것도 좋겠네요 :)

EastHShin commented on src/main/java/roomescape/presentation/PaymentClientController.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.List;
+
+@Component
+public class PaymentClientController {
```
>PaymentClientXXX 의 적절한 패키지 위치는 어디인가.
처음엔 결제 서비스를 제공한다고 생각해 당연히 service 패키지에 위치시켰습니다만, 몇가지 이유로 presentation 패키지로 이동시켰습니다.
service에는 방탈출과 관련된 도메인 로직이 들어있는데, 결제가 방탈출 로직과 관련 없다고 생각했음.
결제 기능은 단순히 요청을 받아서 응답을 반환한다고 생각해 컨트롤러가 있는 presentation에 위치하는게 적절하다고 생각했음.

결제기능은 단순히 요청을 받아서 응답을 반환하는 과정만 있다고 생각하지 않아요. 결제 기능까지 가기 위해 결제 정보와 같은 도메인 클래스가 필요할 수도 있고, 결제 전까지 이루어져야 하는 선결조건이 있을수도 있습니다. 또한, 결제는 비즈니스 로직의 흐름 중 하나이죠.
결제 서비스라는 것을 두고 외부 API를 호출할 수도 있고, 내부적으로 처리할 수도 있습니다. 그렇다면 인터페이스를 두고 이를 구현한 클래스를 두면 될 것이라고 생각되네요.
또한, 결제는 외부 API 를 호출하는 것이기 때문에 만들어주신 패키지중에서는 infrasturcture 가 가장 어울리지 않나 생각되어요.

Astro-Yu replied:
제가 결제 시스템을 단순히 "결제" 라고만 생각한 것 같네요. 말씀하시대로, 결제 여부는 도메인 로직이 결정되는 흐름 속에 존재하는 것인데 따로 생각할 필요가 없을 것 같습니다. Infrasturcture 안에 두고, 결제 Service를 따로 만드는 방식으로 리팩토링 하겠습니다.

EastHShin reviewed (changes requested):
안녕하세요 아마~

다시 뵈어 반갑습니다!
몇가지 코멘트 남겨두었으니 확인 부탁드려요 :)

Astro-Yu committed:
docs: README 수정

Astro-Yu committed:
Timeout 시간 변경

Astro-Yu committed:
style: 코드 스타일 정렬, 사용하지 않는 import 제거

Astro-Yu committed:
refactor: responseErrorHandler 분리

Astro-Yu committed:
refactor: 불필요한 예외 핸들링 제거

Astro-Yu committed:
refactor: 불용 클래스 삭제

Astro-Yu committed:
refactor: 패키지 이동

Astro-Yu committed:
refactor: 새 서비스 생성

PaymentClientController를  PaymentClient로 변경 후 Infra로 이동함.
PaymentService에서 해당 기능을 불러옴

Astro-Yu committed:
refactor: 새 서비스 생성

PaymentClientController를  PaymentClient로 변경 후 Infra로 이동함.
PaymentService에서 해당 기능을 불러옴

Astro-Yu committed:
Revert "refactor: 새 서비스 생성"

This reverts commit bb3b800f11cf94a5e6412a2c76d22c0214c4201e.

Astro-Yu committed:
refactor: 상수 네이밍 변경

Astro-Yu committed:
Merge remote-tracking branch 'origin/step1' into step1

Astro-Yu committed:
refactor: 상수 yml 파일로 분리

Astro-Yu committed:
refactor: ConfigurationProperties 설정 분리

Astro-Yu reviewed (commented):
안녕하세요 이스트 리뷰가 늦어 죄송해요 ㅠㅠ
새로운 변경커밋입니다. [링크](https://github.com/woowacourse/spring-roomescape-payment/pull/197/files/9ee9f1f4da158067ac6e7f550951e135601f55c6..42104908114df3b261f3fa6f93db37b837ced54d)
### 주요 변경사항.
- `PaymentService` 생성, `PaymentClient`는 `Infrastructure.payment` 패키지로 분리
- Client 설정에 사용되는 각종 상수를 yml 파일에 분리 후, `@ConfigurationProperties`를 도입.
- `PaymentClient`의 에러 핸들링을 `paymentResponseErrorHandler`로 분리

Astro-Yu committed:
style: 스타일 정렬

EastHShin commented on src/main/java/roomescape/infrastructure/payment/PaymentClient.java:
```diff
@@ -0,0 +1,31 @@
+                .body(paymentRequest)
+                .retrieve()
+                .onStatus(status ->
+                                status.is4xxClientError() || status.is5xxServerError(),
+                        (request, response) -> paymentResponseErrorHandler.handleError(response,
+                                response.getStatusCode()))
```
handleError 로직에서 reponse를 통해 statusCode를 꺼낼 수 있는데요~
그렇다면 onStatus 로직에서 statusCode 를 확인하는 과정은 필요없이 handler 만 넘겨주어도 되지 않을까요?

Astro-Yu replied:
말씀해주신 방식대로 수정했습니다. 핸들러를 분리하니, 예외 처리 부분이 줄어들어서 가독성이 올라갔네요!

EastHShin commented on src/main/resources/static/js/user-reservation.js:
```diff
@@ -167,30 +183,92 @@ function onReservationButtonClick() {
-                alert("An error occurred while making the reservation.");
-                console.error(error);
-            });
+        // fetch('/api/reservations', {
+        //     method: 'POST',
+        //     headers: {
+        //         'Content-Type': 'application/json',
+        //     },
+        //     body: JSON.stringify(reservationData)
+        // })
+        //     .then(response => {
+        //         if (!response.ok) throw new Error('Reservation failed');
+        //         return response.json();
+        //     })
+        //     .then(data => {
+        //         alert("Reservation successful!");
+        //         location.reload();
+        //     })
+        //     .catch(error => {
+        //         alert("An error occurred while making the reservation.");
+        //         console.error(error);
+        //     });
```
사용하지 않아 주석처리 된 코드는 제거해주세요~

EastHShin commented on src/main/java/roomescape/presentation/ReservationController.java:
```diff
@@ -12,28 +12,37 @@
             @Valid @RequestBody ReservationCreateRequest request) {
-        ReservationResponse reservationResponse = reservationService.createReservation(
-                memberId, request.timeId(), request.themeId(), request.date()
+        PaymentRequest paymentRequest = new PaymentRequest(request.amount(), request.paymentKey(), request.orderId());
+        PaymentInfo paymentInfo = paymentService.createPaymentInfo(paymentRequest);
+
+        ReservationResponse reservationResponse = reservationService.createReservationForMember(
+                memberId, request.timeId(), request.themeId(), request.date(), paymentInfo
         );
```
컨트롤러에서 결제 및 예약 로직이 분리되어 처리되고 있네요~
예약 서비스에서 결제까지 처리해줄 순 없을까요? 컨트롤러에서 이 로직들을 분리해서 처리하게 되면 어떤 문제가 생길 수 있을까요?

Astro-Yu replied:
### 분리했던 이유
어차피 결제 로직은 외부 API라 하나의 트랜잭션에서 관리할 필요가 없다고 생각했습니다. 알아보니 트랜잭션 내에서 외부 api를 사용하는 것도 하나의 커넥션을 오래 점유하여 성능 저하의 문제가 있다고 합니다.
### 해결책
물론 **예약이 반드시 선행되어야 결제가 되는 것은 확실한 도메인 로직이 맞습니다.** 따라서 예약 서비스 안에서 실행하되, 트랜잭션 범위 바깥에서  관리해보도록 하겠습니다.

EastHShin reviewed (changes requested):
안녕하세요 아마

피드백 반영 잘 해주셨네요~
추가 코멘트 남겨두었으니 확인 부탁드려요 :)

Astro-Yu committed:
style: 주석 삭제

Astro-Yu committed:
refactor: ErrorHandler 분리

Astro-Yu committed:
refactor: 예약 생성 로직에 결제 로직 통합

Astro-Yu committed:
refactor: readTimeout 시간 단축

Astro-Yu reviewed (commented):
안녕하세요 이스트! 리뷰 반영해봤습니다. 부족한 부분이 있다면 말씀해주세요.
커밋 변경사항: [링크](https://github.com/woowacourse/spring-roomescape-payment/pull/197/files/9ee9f1f4da158067ac6e7f550951e135601f55c6..5023ee573d88e415e28f7d4bea2fed65d32a35e8)
### 주요 변경점
- PaymentClient에서 errorHandler만 받도록 변경
- 예약 생성과 결제 로직 통합
- readTimeout 시간 단축

Astro-Yu committed:
test: 테스트 수정

EastHShin reviewed (approved):
안녕하세요 아마

미션 요구사항은 모두 충족해 마무리하도록 하겠습니다~
다음 단계에서 뵙겠습니다 :)
